### DISCUSSION TITLE
Performance Improvement Techniques

### DISCUSSION THREAD

**Discussion by sheldonneuberger-sc** [cirruslabs/tart#331] [General]
I'd like to know what people have done or are trying to improve VM performance across any of cpu, disk, network. All ideas/speculation welcome.

Testing on an m1 studio ultra, I'm seeing a large cpu-bound build 20% slower in guest vs host, and another very large build is 40% slower (I suspect it's disk io bound, not sure).

I tried changing VZDiskImageSynchronizationMode to 'none' (it defaults to full), but it didn't have an effect on perf.

Another idea I had was to create a ramdisk and then keep the disk image on there, but at best that would be an 80GB ramdisk, and that's too small for some of my builds.

I assume apple will eventually add the ability to directly attach a physical disk to a VM, which might have much better io perf, but so far that doesn't exist.

**Comment by fkorotkov** [comment]
Hey @sheldonneuberger-sc, could you please tell a little more about your testing scenario? Are you creating a VM that have the same CPU/Memory amount as you M1 Ultra Studio?

What kind of project are you running? Are you mounting it via `tart run --dir` or you clone inside the VM?

**Comment by sheldonneuberger-sc** [comment]
The VM has the same CPU count as the host, memory is not the same but plenty high enough. I'm not doing any --dir mounts. The very large build I'm testing takes about an hour.

**Comment by fkorotkov** [comment]
Here is a [very insightful talk](https://stream.lib.utah.edu/index.php?c=details&id=13535) where at `11:15` Howard is touching the disk performance. Due to usage of the disk image the performance is degraded comparing to the host.

But it does seem like performance of the mounted folders should be on a par with the host. A workaround can be to mount a folder and run your I/O intensive build inside this shared fodler.

**Reply by vj-sc** [reply]
Thanks for sharing the talk @fkorotkov . With a --dir mount, have you ever noticed any file system oddities ? A build that I run consistently fails complaining about a couple of missing files even though the files are present. The files it complains as missing changes from run to run. This error doesn't occur if i don't use the --dir mount. The host is on OS version 13.1 and VM is on version 13.0.

**Reply by fkorotkov** [reply]
Unfortunately, we don't actively use `--dir` ourselves so can't weight on that. But hopefully someone from the community might help. In our use case we just run scripts inside the VM which do fresh Git clones and run CI related scripts.

**Reply by DominikMoreira** [reply]
Hi @vj-sc, I actually came across the exact same thing. With a --dir mount the performance should be on par. But as soon as I run my builds with this setup, it complains about missing files. That's because I set a specific DerivedData-Path to the host. You can fix this by keeping DerivedData in the guest, but then you will lack performance because of I/O tasks (I guess). 

Did you solve this issue? If so, would you mind, sharing it? I would love to take advantage of the near native performance, but for big builds, I guess it will not work.
